src/rfc9457/__init__.py

```diff
@@ -53,35 +58,36 @@ def __repr__(self: t.Self) -> str:
 
     @property
     def type(self: t.Self) -> str:
-        type_ = error_class_to_type(self)
-        return self._type if self._type else type_
+        return self._type if self._type else error_class_to_type(self)
 
-    def marshal(self: t.Self, type_base_url: str | None = None, *, strip_debug: bool = False) -> dict[str, t.Any]:
+    def marshal(
+        self: t.Self,
+        type_base_url: str | None = None,
+        *,
+        uri: str = "",
+        strip_debug: bool = False,
+    ) -> dict[str, t.Any]:
         """Generate a JSON compatible representation.
 
         Args:
         ----
-            type_base_url: If provided prepend to the type to generate a full url.
+            type_base_url: If provided prepend to the type to generate a full url. (DEPRECATED; use uri instead)
+            uri: URI / URI template to use as the type; will substitute '{status}', '{title}', and '{type}'
             strip_debug: If true, remove anything that is not title/type.
         """
-        type_ = f"{type_base_url or ''}{self.type}"
-        ret = {
-            "type": type_,
+        typ = self.type
+        if type_base_url
+            typ = f"{type_base_url or ''}{self.type}"
+            warn("Using deprecated parameter 'type_base_url'", DeprecationWarning)
+        typ = uri.replace("{status}", self.status).replace("{title}", self.title).replace("{type}", self.type) or self.type
+        return {
+            "type": typ,
             "title": self.title,
             "status": self.status,
+            **{k: v for k, v in self.extras.items() if k in self.__mandatory__ or not strip_debug},
```

Comment:
these two feel less readable than previous, a few more mental hoops to jump through to work out what is being included when, but if the outcome is the same (maybe with a comment im not averse)

Reply:
I found that line more readable, but the one below it did feel awkward. I added comments.